RDKCOM-5547: RDKBNETWOR-84 DSLite feature integration with WanManager (Wg network11 development)

devinmungan opened the pull request:
Implement stateful DS-Lite V2 in Wan Manager

Summary
This PR integrates the new generation of DS-Lite into wan-manager and
dhcp-manager. Unlike the previous stateless implementation, this version
is stateful and fully integrated into the Wan Manager state machine.

This implementation moves significant logic (including service_dslite.c
and data model handling) out of Utopia and ccsp-p-and-m into
wan-manager, while maintaining backward compatibility for non-volatile
storage keys.

Technical Implementation
* Architecture Migration:
    * Migrated Device.DSLite TR-181 data model handling from
      ccsp-p-and-m to wan-manager.
    * Moved service_dslite.c logic and non-volatile configuration
      storage from Utopia to wan-manager.
    * Backward Compatibility: Retained existing syscfg keys for
      non-volatile storage to ensure seamless upgrades.

* Data Model & Mapping:
    * InterfaceSetting entries are now maintained in a separate linked
      list.
    * Introduced #define PSM_WANMANAGER_IF_VIRIF_DSLITE_PATH to map
      InterfaceSetting entries to their corresponding Virtual Interfaces
      using the Alias field.

* DHCP Integration:
    * DHCPv6 Option 64 is now received by dhcp-manager and propagated to
      wan-manager, where it is stored in the mapped DSLite entry.

* State Machine:
    * Extended the Virtual Interface state machine to explicitly support
      DS-Lite transition states.

New Features

1. DNS TTL Tracking & Reliability
Added logic to handle endpoint changes and failures gracefully:
* TTL Tracking: Tracks DNS TTL and resolution timestamps for endpoint
  FQDNs.
* Smart Re-resolution: Periodically checks TTL in the DSLITE_ACTIVE
  state. Re-resolves endpoint when TTL expires, but only restarts the
  tunnel if the resolved address has changed.
* Failure Handling: Tears down the tunnel if DNS re-resolution fails to
  prevent stale endpoint usage.
* Retry Logic: Implemented automatic retry every 5 seconds after tunnel
  setup failures (Note: Future improvement planned to make this
  configurable via DM parameter).
* Optimization: Added address caching to avoid redundant DNS queries.

2. Dual Stack Provisioning Support
Introduced specific support for dual-stack provisioning scenarios where
standard IPv6-only DS-Lite is not sufficient.
* Flag: FEATURE_DSLITE_V2_DUALSTACK_SUPPORT (Distro feature:
  dslite_v2_dualstack_support).
* Behavior: If the DS-Lite tunnel fails (e.g., FQDN resolution
  failure), the device will gracefully fallback to the IPv4 address on
  the WAN interface.

Build Configuration
This feature is gated by the following preprocessor macros:
* FEATURE_DSLITE_V2 (Distro feature: dslite_v2)
* FEATURE_DSLITE_V2_DUALSTACK_SUPPORT (Distro feature:
  dslite_v2_dualstack_support)

devinmungan committed:
RDKBNETWOR-84: DSLite feature integration with WanManager (Wg network11 development)

* RDKBNETWOR-86 : Move Device.DSLite. datamodel implementation to WAN Manager

* DSLite Data Model Migration to WAN Manager

Signed-off-by: Devin Mungan <devin.mungan@infosys.com>

govindaraja-nb committed:
RDKBNETWOR-84: DSLite feature integration with WanManager (Wg network11 development)

* RDKBNETWOR-87: DSLite tunnel interface management

* Moved service_dslite from Utopia to WanManager.

Signed-off-by: Govindaraja N B <gnb.ext@libertyglobal.com>

devinmungan committed:
RDKBNETWOR-84: DSLite feature integration with WanManager (Wg network11 development)

* RDKBNETWOR-85 : DHCP Manager changes for option 64

* parse and import endpoint name data coming from dhcp-manager

Signed-off-by: Devin Mungan <devin.mungan@infosys.com>

devinmungan committed:
RDKBNETWOR-84: DSLite feature integration with WanManager (Wg network11 development)

* RDKBNETWOR-88 : WAN Manager state machine for DSLite

* Interface state machine is extended to cover DS-Lite mode as well

Signed-off-by: Devin Mungan <devin.mungan@infosys.com>

devinmungan committed:
RDKBNETWOR-84: DSLite feature integration with WanManager (Wg network11 development)

* RDKBNETWOR-89 : WAN Manager DSLite failure/failover handling

Summary
Added DNS TTL tracking with periodic re-resolution and retry logic
to handle endpoint changes and failures gracefully.

Key Changes

- Track DNS TTL and resolution timestamp for endpoint FQDNs
- Periodically check TTL expiration in DSLITE_ACTIVE state
- Re-resolve endpoint when TTL expires; only restart tunnel if address changed
- Tear down tunnel if DNS re-resolution fails (prevents stale endpoint usage)
- Automatic retry every 5 seconds after tunnel setup failures (TODO is adding a DM parameter for it)
- Optimize DNS queries with address caching to avoid redundant resolutions

Signed-off-by: Devin Mungan <devin.mungan@infosys.com>

devinmungan committed:
RDKBNETWOR-84: DSLite feature integration with WanManager (Wg network11 development)

* RDKBNETWOR-89 : WAN Manager DSLite failure/failover handling

Add Dual-Stack device-mode awareness to the DS-Lite feature.
By design, DS-Lite is enabled only when the CPE operates in IPv6-only mode.
For platforms that intentionally want DS-Lite in Dual-Stack mode, enable the distro feature:
dslite_v2_dualstack_support

Signed-off-by: Devin Mungan <devin.mungan@infosys.com>

devinmungan committed:
RDKBNETWOR-84: DSLite feature integration with WanManager (Wg network11 development)

* Enhance error handling and update the state-machine flow
  to align with the latest design agreed with the Tiger Team.

sichevoi commented on config/RdkWanManager_v2.xml:
```diff
@@ -1436,6 +1440,142 @@
+                        <writable>true</writable>
+                    </parameter>
+                    <parameter>
+                        <name>TunnelV4Addr</name>
```
I couldn't find this one as part of the latest TR-181 https://cwmp-data-models.broadband-forum.org/tr-181-2-20-1-cwmp.html - is this a new vendor-specific parameter? Need to rename to X_RDKCENTRAL-COM_TunnelV4Addr , then.

sichevoi commented on source/TR-181/middle_layer_src/wanmgr_dml_iface_v2_apis.c:
```diff
@@ -1524,7 +1524,21 @@ BOOL WanVirtualIf_GetParamBoolValue(ANSC_HANDLE hInsContext, char* ParamName, BO
+                {
+                    *pBool = entry->CurrCfg.Enable;
+                    WanMgr_GetDSLiteData_release();
+                }
```
suggestion: log the failure.

sichevoi commented on source/TR-181/middle_layer_src/wanmgr_rdkbus_apis.c:
```diff
@@ -260,7 +261,15 @@ int get_Virtual_Interface_FromPSM(ULONG instancenum, ULONG virtInsNum ,DML_VIRTU
+#endif
 
+#ifdef FEATURE_DSLITE_V2
+    _ansc_memset(param_name, 0, sizeof(param_name));
```
I thought RDK-B was moving away from ANSC framework.

sichevoi reviewed (commented):
I'll continue later this week.

sichevoi commented on source/TR-181/include/wanmgr_dml.h:
```diff
@@ -61,6 +61,16 @@ typedef enum _DEVICE_NETWORKING_MODE_
 } DEVICE_NETWORKING_MODE;
 
+typedef  enum
+_DML_WAN_DEVICE_MODE
```
Got it!

sichevoi commented on source/TR-181/middle_layer_src/wanmgr_dml_dslite_apis.h:
```diff
@@ -0,0 +1,72 @@
+BOOL DSLite_SetParamBoolValue(ANSC_HANDLE hInsContext, char *ParamName, BOOL bValue);
+BOOL DSLite_GetParamUlongValue(ANSC_HANDLE hInsContext, char *ParamName, ULONG *pUlong);
+
+ULONG InterfaceSetting4_GetEntryCount(ANSC_HANDLE hInsContext);
```
Let it be :)

sichevoi commented on source/WanManager/wanmgr_data.c:
```diff
@@ -1114,3 +1321,42 @@ DML_VIRTUAL_IFACE* WanMgr_GetActiveVirtIfData_locked(void)
+                }
+            }
+        }
+        WanMgrDml_GetIfaceData_release(NULL);
```
suggestion: I'd either reworked it to use WanMgr_GetIfaceData_locked() (and then iterate, so you'll have to release the mutex, will be a bit more costly) - or use pthread_mutex_unlock() directly here. It's a strange implementation for WanMgrDml_GetIfaceData_release() - the argument isn't used, if it were to be reimplemented to match the design, there might be a problem here.

sichevoi reviewed (commented):
I've finished this part.

devinmungan commented on source/WanManager/wanmgr_sysevents.c:
```diff
@@ -1512,3 +1512,43 @@ void wanmgr_get_wan_interface(char *wanInterface)
+
+    if (pid > 0)
+    {
+        sprintf(cmdline_buf, "/proc/%d/cmdline", pid);
```
probably copied from utopia. @govindaraja-nb need your input pls

govindaraja-nb replied:
Yes, this logic was copied from Utopia (service_dslite).
So instead of keeping this logic, I updated it to simply restart Zebra. Also moved all service restarts into helper WanMgr_Dslite_RestartServices().

devinmungan commented on source/WanManager/wanmgr_t2_telemetry.c:
```diff
@@ -176,6 +176,25 @@ ANSC_STATUS wanmgr_process_T2_telemetry_event(WanMgr_Telemetry_Marker_t *Marker)
+            wanmgr_telemetry_append_key_value(WANMGR_T2_SELECTION_STATUS_STRING,(pIntf->Selection.Status == WAN_IFACE_ACTIVE) ? "Active" : (pIntf->Selection.Status == WAN_IFACE_VALIDATING) ? "Validating" : (pIntf->Selection.Status == WAN_IFACE_SELECTED ) ? "Selected" : "Standby");
+            break;
+
+        case WAN_ERROR_DSLITE_STATUS_FAILED:
```
let's add it

devinmungan committed:
Addressing review comments

- change dhcpv6 & dml related AFTR FQDN naming to aftr instead of endpointName
  keep internal naming as "endpointName" since internals also will have endpointAddr
  and these parameters are already under DSLite context therefore both function
  and struct namings suggest endPointName is a DSLite related field.
- change TunnelV4Addr to X_RDKCENTRAL-COM_TunnelV4Addr as it's not a standard TR-181 DM
- convert ANSC to generic
- link libevent to wanmgr only if DSLITE_SUPPORT flag exists
- add telemetry event for the case WAN_ERROR_DSLITE_STATUS_FAILED

devinmungan commented on source/TR-181/include/wanmgr_dml.h:
```diff
@@ -367,6 +392,9 @@ typedef struct _WANMGR_IPV6_DATA
    bool domainNameAssigned;     /**< Have we been assigned dns server addresses ? */
    #endif
+#ifdef FEATURE_DSLITE_V2
+   char endpointName[BUFLEN_256];
```
I've changed naming to aftr for DHCPv6 and DML related fields. WanMgr internals will still use endpointName and endpointAddr since we have 2 ways of representing the aftr (this is not the case for dhcpv6, only fqdn is present) and both function and structs namings clearly indicate the contexts is DSLite

govindaraja-nb committed:
Improve DSLite service restart handling

- add helper to restart required services during DSLite up/down
- remove wanmgr_restart_zebra() and move zebra restart into the helper function.

Signed-off-by: Govindaraja N B <gnb.ext@libertyglobal.com>

devinmungan commented on source/WanManager/wanmgr_dslite.c:
```diff
@@ -0,0 +1,1277 @@
+    }
+
+    /* Check if any user-configurable fields changed (excludes runtime fields like Status, AddrInUse, etc.) */
+    if (prev->Enable != curr->Enable ||
+        strcmp(prev->Alias, curr->Alias) != 0 ||
+        prev->Mode != curr->Mode ||
+        prev->Type != curr->Type ||
+        strcmp(prev->EndpointName, curr->EndpointName) != 0 ||
+        strcmp(prev->EndpointAddr, curr->EndpointAddr) != 0 ||
+        strcmp(prev->TunnelV4Addr, curr->TunnelV4Addr) != 0 ||
+        prev->MssClampingEnable != curr->MssClampingEnable ||
+        prev->TcpMss != curr->TcpMss ||
+        prev->Ipv6FragEnable != curr->Ipv6FragEnable)
```
check

govindaraja-nb committed:
simplify DSLite tunnel teardown logic

- Remove address retrieval and delete the tunnel directly using the tunnel interface name.

Signed-off-by: Govindaraja N B <gnb.ext@libertyglobal.com>

S-Parthiban-Selvaraj commented on config/RdkWanManager_v2.xml:
```diff
@@ -414,7 +414,11 @@
                                     <name>EnableDSLite</name>
                                     <type>boolean</type>
                                     <syntax>bool</syntax>
+<?ifdef FEATURE_DSLITE_V2?>
```
Are we disabling runtime change of the dslite feature?
Only PSM config works?

devinmungan replied:
no, we only disabled "EnableDSLite" usage which was not being used anyway. dslite will now be controlled from it's TR-181 dedicated paths
Device.DSLite.Enable
Device.DSLite.InterfaceSetting.x.Enable

S-Parthiban-Selvaraj commented on source/TR-181/middle_layer_src/wanmgr_dml_iface_v2_apis.c:
```diff
@@ -1524,7 +1524,21 @@ BOOL WanVirtualIf_GetParamBoolValue(ANSC_HANDLE hInsContext, char* ParamName, BO
+            *pBool = FALSE;
+            if (p_VirtIf->DSLite.Path[0] != '\0')
+            {
+                DML_DSLITE_LIST *entry = WanMgr_getDSLiteEntryByAlias_locked(p_VirtIf->DSLite.Path);
```
Ideally in the path , we use the full DML itself, 
it would be easier to directly set the DML name instead of finding with the alias

S-Parthiban-Selvaraj commented on source/WanManager/Makefile.am:
```diff
@@ -48,3 +48,8 @@ wanmanager_SOURCES += wanmgr_dhcp_event_handler.c
+
+if DSLITE_SUPPORT
+wanmanager_SOURCES += wanmgr_dslite.c
+wanmanager_LDADD += -levent
```
is it a new lib ?

devinmungan replied:
Yes, libevent is added to utilize libevent's **evdns** module to resolve AFTR address.

S-Parthiban-Selvaraj commented on source/WanManager/wanmgr_dslite.c:
```diff
@@ -0,0 +1,1246 @@
+        DML_DSLITE_CONFIG cfg;
+        DSLITE_SET_DEFAULTVALUE(&cfg);  /* start from known defaults */
+
+        snprintf(key, sizeof(key), "dslite_active_%d", insNum);
```
Can this be migrated to PSM ?

S-Parthiban-Selvaraj commented on source/TR-181/middle_layer_src/wanmgr_dml_dslite_apis.c:
```diff
@@ -0,0 +1,659 @@
+#include "wanmgr_rdkbus_apis.h"
+#include "wanmgr_net_utils.h"
+#include "wanmgr_data.h"
+#if defined (_XB6_PRODUCT_REQ_) || defined (_CBR2_PRODUCT_REQ_) || defined(_PLATFORM_RASPBERRYPI_)
```
Do we need this product flags?

devinmungan replied:
it's a copy-paste result. those flags were present in _wanmgr_dml_iface_apis.c_ and _wanmgr_dml_iface_v2_apis.c_

S-Parthiban-Selvaraj commented on source/TR-181/middle_layer_src/wanmgr_dml_dslite_apis.c:
```diff
@@ -0,0 +1,659 @@
+#ifndef FEATURE_DSLITE_V2_DUALSTACK_SUPPORT
+            UINT deviceMode = 0;
+
+            WanMgr_SysCfgGetUint("last_erouter_mode", &deviceMode);
```
Please add a TODO to use the WanManager PSM later.

devinmungan replied:
sure

S-Parthiban-Selvaraj commented on source/WanManager/wanmgr_dslite.c:
```diff
@@ -0,0 +1,1246 @@
+                   __FUNCTION__, pVirtIf->Name, elapsed, cfg->DnsTtl));
+
+    // Get the endpoint FQDN to re-resolve
+    if (cfg->Mode == DSLITE_ENDPOINT_DHCPV6)
```
both if and else looks same

devinmungan replied:
I'll change it

S-Parthiban-Selvaraj commented on source/WanManager/wanmgr_interface_sm.c:
```diff
@@ -1699,6 +1730,90 @@ static int wan_tearDownIPv6(WanMgr_IfaceSM_Controller_t * pWanIfaceCtrl)
+    if (p_VirtIf->IP.Mode == DML_WAN_IP_MODE_IPV6_ONLY)
+    {
+        /* IPv6-only mode: disable LAN-to-WAN IPv4 forwarding */
+        if (sysctl_iface_set("/proc/sys/net/ipv4/ip_forward", NULL, "0") != 0)
```
ip_forward
we have to make sure if this is reset, if we move to a backup/ different inetrface

devinmungan replied:
**wan_setUpIPv4** function should already handle this case if the selected interface supports ipv4. Or do you mean  **wan_setUpIPv4** for newly selected interface may run before this reset?

S-Parthiban-Selvaraj commented on source/WanManager/wanmgr_interface_sm.c:
```diff
@@ -4277,6 +4572,78 @@ static eWanState_t wan_state_map_active(WanMgr_IfaceSM_Controller_t* pWanIfaceCt
+    WanMgr_MonitorDhcpApps(pWanIfaceCtrl);
+
+    WanMgr_CheckDefaultRA(p_VirtIf);
+#if defined(FEATURE_IPOE_HEALTH_CHECK) && defined(IPOE_HEALTH_CHECK_LAN_SYNC_SUPPORT)
```
Lets remove this code, I am removing this code from the SM in another PR

devinmungan replied:
do you mean this block?

```
#if defined(FEATURE_IPOE_HEALTH_CHECK) && defined(IPOE_HEALTH_CHECK_LAN_SYNC_SUPPORT)
    if (lanState == LAN_STATE_STOPPED)
    {
        WanMgr_SendMsgTo_ConnectivityCheck(pWanIfaceCtrl, CONNECTION_MSG_IPV6, FALSE);
        lanState = LAN_STATE_RESET;
    }
    else if (lanState == LAN_STATE_STARTED)
    {
        WanMgr_SendMsgTo_ConnectivityCheck(pWanIfaceCtrl, CONNECTION_MSG_IPV6, TRUE);
        lanState = LAN_STATE_RESET;
    }
#endif
```

S-Parthiban-Selvaraj commented on source/WanManager/wanmgr_interface_sm.c:
```diff
@@ -3075,6 +3204,102 @@ static eWanState_t wan_transition_map_down(WanMgr_IfaceSM_Controller_t* pWanIfac
+        return WAN_STATE_IPV6_LEASED;
+    }
+
+#ifdef FEATURE_DSLITE_V2_DUALSTACK_SUPPORT
```
Do we need this MACRO ?
can we control this with DHCP interface configuration Ipv6 only ?

devinmungan replied:
a customer may support dualstack, but not in DSLite mode. In that case, we either need this flag, or trust the customer to set the correct config before/after enabling/disabling DSLite.

Even better, maybe **p_VirtIf->IP.Mode** can directly be used. What do you think? (This may require some additional changes)

All these aside, DSLite's purpose is to operate when ipv4 is not possible, therefore dualstack support itself doesn't make much sense to me but there was a request about it.

Anyway, I'll think on this a bit.

devinmungan committed:
Addressing further review comments on DSLITE_V2 implementation

guto86 commented on source/WanManager/wanmgr_data.c:
```diff
@@ -1131,3 +1339,42 @@ DML_VIRTUAL_IFACE* WanMgr_GetActiveVirtIfData_locked(void)
+                {
+                    if(!strcmp(Alias,virIface->DSLite.Path))
+                    {
+                        return virIface;
```
mutex?

devinmungan replied:
you mean unlock is missing? If so, The caller is responsible for unlocking the mutex, consistent with the convention used by other _locked helpers.
